[FLINK-21259] Add Failing state for DeclarativeScheduler

rmetzger opened the pull request:
This PR is based on https://github.com/apache/flink/pull/14879

## What is the purpose of the change

[Declarative Scheduler](https://cwiki.apache.org/confluence/display/FLINK/FLIP-160%3A+Declarative+Scheduler) consists of a number of internal states. 

Note that this change is currently not usable as-is, as the other parts of declarative scheduler are not merged yet (See for the prototype this PR is based on: https://github.com/tillrohrmann/flink/tree/declarative-scheduler) 


## Verifying this change

- The change is adding unit tests.
- Note that integration tests for the declarative scheduler will cover additional functionality.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

Will be handled in a separate PR.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StateWithExecutionGraph.java:
```diff
@@ -138,7 +138,9 @@ public JobStatus getJobStatus() {
         executionGraph.suspend(cause);
-        Preconditions.checkState(executionGraph.getState() == JobStatus.SUSPENDED);
+        Preconditions.checkState(
+                executionGraph.getState() == JobStatus.SUSPENDED
+                        || executionGraph.getState().isTerminalState());
```
Why is this change necessary?

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/FailingTest.java:
```diff
@@ -0,0 +1,151 @@
+    @Test
+    public void testFailJobOnEnter() throws Exception {
+        try (MockFailingContext ctx = new MockFailingContext()) {
+            Failing failing = createFailingState(ctx);
```
I guess that this creates an `ExecutionGraph` which directly goes to `FAILED` when calling `ExecutionGraph.fail` because there are no deployed/running `Executions`. I think this is a good example what a proper testing abstraction of the `ExecutionGraph` could help us with. We would no longer bind the logic of the tests to the underlying `ExecutionGraph` implementation.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/FailingTest.java:
```diff
@@ -0,0 +1,151 @@
+    @Test
+    public void testIgnoreGlobalFailure() throws Exception {
+        try (MockFailingContext ctx = new MockFailingContext()) {
+            Failing failing = createFailingState(ctx);
```
I guess `onEnter()` is missing here. By not sticking to the contract, we are testing scenarios which should not happen in "production".

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/FailingTest.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Test
+    public void testTransitionToCancelOnCancel() throws Exception {
```
`testTransitionToCancelingOnCancel`

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/FailingTest.java:
```diff
@@ -0,0 +1,151 @@
+    @Test
+    public void testTransitionToCancelOnCancel() throws Exception {
+        try (MockFailingContext ctx = new MockFailingContext()) {
+            Failing failing = createFailingState(ctx);
```
`onEnter()` is missing

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/FailingTest.java:
```diff
@@ -0,0 +1,151 @@
+            Failing failing = createFailingState(ctx);
+            ctx.setExpectFinished(
+                    archivedExecutionGraph ->
+                            assertThat(archivedExecutionGraph.getState(), is(JobStatus.FAILED)));
```
The final `JobStatus` should be `SUSPENDED`.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @rmetzger. The changes go in a good direction. I had a couple of comments concerning the added tests. Please take a look.

rmetzger commented:
Rebased to latest master, will address the comments tomorrow (ideally after the Canceling PR is merged, so that I can share the MockExecutionGraph).

rmetzger commented:
I addressed all your comments 
(note: "the address feedback" commit contains 98% of my changes. I messed up an interactive rebase, the "add tests" commit contains a move of the MockExecutionEnvironment out of the CancellingTest to a top level class).

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/FailingTest.java:
```diff
@@ -0,0 +1,178 @@
+                                    new RuntimeException()));
+            failing.updateTaskExecutionState(update);
+            ctx.assertNoStateTransition();
+            assertThat(meg.isFailGlobalCalled(), is(true));
```
I think we are doing this because `updateTaskExecutionState` fails with a `NPE`.

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @rmetzger. I had a few more comments. Please take a look.

rmetzger commented:
Thanks for your review. I addressed your comments!

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/MockExecutionGraph.java:
```diff
@@ -69,15 +73,22 @@ public MockExecutionGraph() throws IOException {
                 (execution, newState) -> {},
                 0L);
         this.setJsonPlan(""); // field must not be null for ArchivedExecutionGraph creation
+        this.attachJobGraph(Collections.emptyList()); // method must be called to initialize
+        // PartitionReleaseStrategy.
+    }
+
+    private static PartitionReleaseStrategy.Factory getPartitionReleaseStrategyFactoryLoader() {
+        Configuration conf = new Configuration();
+        conf.set(PARTITION_RELEASE_DURING_JOB_EXECUTION, false);
+        return PartitionReleaseStrategyFactoryLoader.loadPartitionReleaseStrategyFactory(conf);
```
Hmm, I strongly suggest to add an interface for testing in the foreseeable future. I think we are seeing the consequences of mocking the `ExecutionGraph` by running into problems originating from setting it up wrongly.

rmetzger replied:
Yes, I agree. This is very fragile and annoying to maintain.

tillrohrmann reviewed (approved):
Thanks for updating this PR @rmetzger. LGTM. +1 for merging once AZP gives green light.

rmetzger commented:
Thanks for the review! I'll merge this now.
